src/openvic-simulation/economy/Producer.hpp

```diff
@@ -0,0 +1,73 @@
+#pragma once
+
+#include <cstddef>
+#include <cstdint>
+
+#include "openvic-simulation/economy/Good.hpp"
+#include "openvic-simulation/economy/ProductionType.hpp"
+#include "openvic-simulation/types/fixed_point/FixedPoint.hpp"
+#include "openvic-simulation/utility/Getters.hpp"
+
+namespace OpenVic {
+	class Producer {
+	public:
+		ProductionType const& PROPERTY(production_type);
+	};
+
+	class Manufacturer : public virtual Producer {
+	public:
+		Good::good_map_t PROPERTY(stockpile);
+	};
+
+	class ArtisanalProducer final : public Manufacturer {
+	public:
+		fixed_point_t PROPERTY(current_production);
+		Good::good_map_t PROPERTY(current_needs);
+	};
+
+	class NonArtisanalProducer : public virtual Producer {
+	public:
+		fixed_point_t PROPERTY(revenue_yesterday);
+		fixed_point_t PROPERTY(output_quantity_yesterday);
+		fixed_point_t PROPERTY(unsold_quantity_yesterday);
+		fixed_point_t PROPERTY(size_multiplier);
+		ordered_map<Pop const*, Pop::pop_size_t> PROPERTY(employees);
+	};
+
+	class FactoryProducer final : public NonArtisanalProducer, public Manufacturer {
+	private:
+		static constexpr size_t days_of_history = 7;
+		int PROPERTY(profit_history_current);
+		std::array<fixed_point_t, days_of_history> PROPERTY(daily_profit_history);
```

Comment:
```suggestion
		static constexpr size_t DAYS_OF_HISTORY = 7;
		size_t PROPERTY(profit_history_current);
		std::array<fixed_point_t, DAYS_OF_HISTORY> PROPERTY(daily_profit_history);
```
- Although we're not always 100% consistent with this, it's useful to give constants all-caps names so that it's clear that they're not mutable variables.
- Let's make `profit_history_current` a `size_t` so it's the same type as `DAYS_OF_HISTORY` which it'll be compared to. This doesn't take up any more space, as there are currently 4 bytes / 32 bits of padding space after `profit_history_current` so that `daily_profit_history` is at an 8 byte / 64 bit aligned position. By default variables in a struct are aligned to `min(8, their size) rounded up to a power of 2` byte positions, with padding space added in between.

Also `profit_history_current` will need to be initialised to 0 by `FactoryProducer`'s constructor, as MSVC doesn't automatically default initialise primitive variables.